### PR TITLE
NCBIClient (from internal)

### PR DESCRIPTION
**Pull request by sygao**
- Add `virtool_cli.ncbi.NCBIClient` and `virtool_cli.ncbi.NCBICache`
	- Add `NCBIClient.fetch_accessions()`,  `NCBIClient.fetch_taxonomy()`
		- fetches and parses in a single method
		- caches data if flag is set
	- Add `NCBIClient.link_accessions_to_taxid()` for use in update functions
	- Add corresponding tests in `test_ncbi_client` and `test_ncbi_cache`
		- run `pytest -m "not ncbi" --disable-socket --allow-unix-socket` to run NCBI database-free tests
- Add Pydantic validation in `virtool_cli.ncbi.models`
	- Add corresponding tests for validation

### CODE REVIEW

File: tests/test_ncbi_client.py [hunk 1]
```diff
@@ -1,91 +1,260 @@
...
+    "accessions",
+    [
+        ["AB017504", "MH200607", "MK431779", "NC_003355"],
+        ["NC_036587", "MT240513", "MT240490"],
+    ],
+)
+class TestClientFetchRawGenbank:
+    @pytest.mark.ncbi
+    @pytest.mark.asyncio
+    async def test_fetch_raw_via_accessions(self, accessions):
```
Comment (igboyes) [inline tests/test_ncbi_client.py:136]:
```suggestion
    async def test_fetch_raw_via_accessions(self, accessions: list[str]):
```

**Review by igboyes — CHANGES_REQUESTED** [review]
* The biggest thing that needs to be fixed here is handling the flaky `400` errors. Reading the response will give a clue as to the reason if you haven't tried it.
  ```py
  except HTTPError as e:
    print(e.read())
  ```
* It seems to me that _genbank_ and _nuccore_ are being used somewhat interchangeably. I think it would be best to use _genbank_ as it will be more familiar to biologists. The only thing that needs to stay `nuccore` is variant in the database enum. Thoughts?
* If you are still having problems with the `NCBI_` environmental variables not getting into the tests, moving the `get` and setting of the `Entrez` attributes to `virtool_cli/__init__.py` should fix the problem. I don't think the code is imported where it is now during testing and thus is not being executed.

File: tests/test_ncbi_cache.py [hunk 2]
```diff
@@ -35,54 +37,73 @@ def test_cache_init(empty_cache_path):
 def test_cache_clear(cache_scratch_path):
     cache = NCBICache(path=cache_scratch_path)
 
+    try:
+        assert next(cache.nuccore.glob("*.json"))
+        assert next(cache.taxonomy.glob("*.json"))
+    except StopIteration as exc:
+        pytest.fail(f"Cache scratch path was not set up correctly: {exc}")
```
[the remark this answers is not in the document]
Comment (sygao) [reply tests/test_ncbi_cache.py:44]: This part of the test only needs to make sure the cache directories have JSON files in them. I don't need to actually do anything with the contents, so this seems more efficient than turning the generator into a list.

File: tests/test_ncbi_client.py [hunk 3]
```diff
@@ -1,91 +1,260 @@
...
+    )
+    async def test_fetch_partially_cached_genbank_records(
+        self, accessions, cache_scratch_path
+    ):
+        client = NCBIClient(cache_scratch_path)
+
+        try:
+            assert next(client.cache.nuccore.glob("*.json"))
+        except StopIteration:
+            pytest.fail("Could not retrieve any records from cache")
```
[the remark this answers is not in the document]
Comment (sygao) [reply tests/test_ncbi_client.py]: Can't be sure if there isn't any sort of system chaff in there unless the environment is 100% fresh. Checking only the relevant filetype is the cleanest way to ignore system chaff.

File: tests/test_ncbi_client.py [hunk 4]
```diff
@@ -1,91 +1,260 @@
...
+async def test_check_spelling(misspelled, expected):
+    await asyncio.sleep(DEFAULT_PAUSE)
+    try:
+        taxon_name = await NCBIClient.check_spelling(name=misspelled)
+    except HTTPError:
+        await asyncio.sleep(SAFETY_PAUSE)
+        taxon_name = await NCBIClient.check_spelling(name=misspelled)
 
-        assert taxon_name.lower() == expected.lower()
+    assert taxon_name.lower() == expected.lower()
```
[the remark this answers is not in the document]
Comment (sygao) [reply tests/test_ncbi_client.py:236]: == has case insensitivity. I'll just modify the input data for now.

File: tests/test_ncbi_cache.py [hunk 5]
```diff
@@ -10,79 +11,98 @@ def empty_cache_path(tmp_path):
...
+    assert cache._nuccore_path.is_dir()
 
-    assert cache.taxonomy.is_dir()
+    assert cache._taxonomy_path.is_dir()
 
-    assert cache.nuccore.exists()
+    assert cache._nuccore_path.exists()
 
-    assert cache.taxonomy.exists()
+    assert cache._taxonomy_path.exists()
```
Comment (igboyes) [inline tests/test_ncbi_cache.py:35]: Don't access private members. Test like:
```py
assert (empty_cache_path / ".cache" / "nuccore").is_dir()
```
Use `is_dir()`. The `exists()` method is less specific.

File: tests/test_ncbi_cache.py [hunk 6]
```diff
@@ -10,79 +11,98 @@ def empty_cache_path(tmp_path):
...
 
 
 def test_cache_clear(cache_scratch_path):
     cache = NCBICache(path=cache_scratch_path)
 
-    cache.clear()
-
-    assert not list(cache.nuccore.glob("*.json"))
+    assert list(cache._nuccore_path.glob("*.json")) != []
+    assert list(cache._taxonomy_path.glob("*.json")) != []
```
Comment (igboyes) [inline tests/test_ncbi_cache.py:42]: Why use glob? The directory should be completely empty of anything.

**Review by igboyes — APPROVED** [review]

**Pull request merged**